crates/sage-cloudpath/src/lib.rs

```diff
@@ -242,6 +242,16 @@ where
 {
     let path = path.as_ref().parse::<CloudPath>()?;
 
+    // return an error in case we are trying to read from a bucket without a key
+    match &path {
```

Comment:
I think this should be moved up into `Cloudpath::read`, and I think would be better as an `if let`:

```rust
if let CloudPath::S3 { key, .. } = &path {

}
```

Reply:
yes.. totally right about the `if let`. I added the check in `read_and_execute` to fail as soon as possible, before doing stuff with tokio.